docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,186 @@ the worker pool has a finite size and even with the full batching algorithm in t
+alternatively, if the transaction was aborted).
+
+#### Coordination Service
+
+We introduce a notion of a coordination service, which agrees on values being relevant or correct at a given timestamp.
+The sequence of values being agreed needs to evolve in a *backwards consistent* manner - that is, one should avoid
+changes that may affect behaviour for decisions made at timestamps before the point where one is making one's update.
+More formally, if we read a value for some timestamp ``TS``, then all future values written must then ensure that
+decisions made at ``TS`` would be done in a way consistent with our initial read.
+
+Values provided are valid up to a specific timestamp, called the validity bound. When a transform is applied, a fresh
+timestamp ``F`` is taken, and the new value will be written with a validity bound ``F + C`` for some constant ``C``.
+Going forward, the behaviour for updates up to the previous bound must be preserved (which means decisions would be the
+same whether other service nodes read the new value or not). Nodes must not make decisions based on values read with a
+validity bound less than their timestamp of interest.
+
+```java
+public interface CoordinationService<T> {
+    Optional<ValueAndBound<T>> getValueForTimestamp(long timestamp);
+
+    CheckAndSetResult<ValueAndBound<T>> tryTransformCurrentValue(Function<ValueAndBound<T>, T> transform);
+
+    // ONLY FOR METRICS AND MONITORING - NOT FOR PRODUCTION DECISIONS.
+    Optional<ValueAndBound<T>> getLastKnownLocalValue();
+}
+```
+
+##### Coordination Service Persistence
+
+The coordination service needs to persist state; we've implemented this as a non-transactional table in the AtlasDB
+key-value service called ``_coordination``. Within this table, each sequence of values being agreed on takes up one row.
+The values themselves are stored as dynamic columns with a var-long dynamic column key.
+
+- The dynamic column value stored with dynamic column key 0 is a ``SequenceAndBound`` which indicates the sequence
+  number of the currently valid value and the validity bound.
+- The values to be agreed upon are stored as dynamic column values. These are stored at dynamic column keys which
+  correspond to fresh timestamps taken at the time they were written. Dynamic column keys are stored as ``VAR_LONG``s.
+
+Conceptually, the state of the table may look as follows:
+
+| Sequence Id | Sequence Number |                              Value |
+|------------:|----------------:|-----------------------------------:|
+|           m |               0 | {"sequence":13007,"bound":5013007} |
+|           m |           10001 |                  <metadata object> |
+|           m |           13007 |        <different metadata object> |
+
+The sequence ID is the same, to reflect that these values are a part of the same sequence; ``m`` is chosen for
+AtlasDB's internal schema metadata in general, and isn't particularly meaningful in and of itself.
```

Comment:
might be worth mentioning either here or in the `Determining Transaction Schema Versions` section that this isn't inherently timestamp aware, and it's up to the implementation for a sequence id `m` to partition the timestamp space etc. Possibly also introducing something along the lines of, adding an abstraction layer (which already exists) such that this is handled for you automatically i.e. you never deal with the coordination service directly etc.